pkg/kubelet/mirror_manager.go

```diff
@@ -0,0 +1,165 @@
+/*
+Copyright 2015 Google Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubelet
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+	"github.com/golang/glog"
+)
+
+// Kubelet discover pod updates from 4 sources: file, http, etcd, and
+// apiserver. Pods from non-apiserver sources are called static pods, and API
+// server is not aware of the existence of static pods. In order to monitor the
+// status of such pods, kubelet create a mirror pod for each static pod via the
+// API server.
+//
+// A mirror pod has the same pod full name (name and namespace) as its static
+// counterpart (albeit different metadata such as UID, etc). By leveraging the
+// fact that kubelet reports the pod status using the pod full name, the status
+// of the mirror pod always reflects the acutal status of the static pod.
+// When a static pod gets deleted, the associated orphaned mirror pods will
+// also be removed.
+//
+// This file includes functions to manage the mirror pods.
+
+type mirrorManager interface {
+	CreateMirrorPod(*api.BoundPod, string) error
+	DeleteMirrorPod(string) error
+}
+
+type basicMirrorManager struct {
+	// mirror pods are stored in the kubelet directly because they need to be
+	// in sync with the internal pods.
+	apiserverClient client.Interface
+}
+
+func newBasicMirrorManager(apiserverClient client.Interface) *basicMirrorManager {
+	return &basicMirrorManager{apiserverClient: apiserverClient}
+}
+
+// Creates a mirror pod.
+func (self *basicMirrorManager) CreateMirrorPod(pod *api.BoundPod, hostname string) error {
+	if self.apiserverClient == nil {
```

Comment:
We need to support standalone kubelets, don't we? Mirror pods are an add-on, which should not affect kubelet's basic functions. Or were you suggesting that I moved the check elsewhere?